File: assets/rt/MutStr.java
package rt;

import base.Bool_0;
import base.False_0;
import base.True_0;

import java.nio.ByteBuffer;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.concurrent.locks.ReentrantLock;
import java.util.stream.Collectors;

public final class MutStr implements Str {
  private ByteBuffer buffer = ByteBuffer.allocateDirect(16);
  private volatile int[] graphemes = null;

  public MutStr(Str str) {
    assert str != null;
    put(str);
  }

  @Override public ByteBuffer utf8() {
    return buffer.slice(0, buffer.position());
  }

  @Override public int[] graphemes() {
    var graphemes = this.graphemes;
    if (graphemes != null) {
      return graphemes;
    }
    graphemes = NativeRuntime.indexString(utf8());
    this.graphemes = graphemes;
    return graphemes;
  }

  @Override public Bool_0 isEmpty$read() {
    return buffer.position() == 0 ? True_0.$self : False_0.$self;
  }

  @Override public MutStr $plus$mut(base.Stringable_0 other$) {
    var other = other$.str$read();
    assert other != null;
    put(other);
    return this;
  }

  @Override public base.Void_0 append$mut(base.Stringable_0 other$) {
    this.$plus$mut(other$);
    return base.Void_0.$self;
  }

  @Override public base.Void_0 clear$mut() {
    buffer = ByteBuffer.allocateDirect(16);
    return base.Void_0.$self;
  }

  @Override public rt.Str str$read() {
    return rt.Str.fromTrustedUtf8(utf8());
  }

  private void put(Str str) {
    graphemes = null;
    var toAdd = str.utf8();
    assert toAdd.position() == 0;
    if (buffer.remaining() < toAdd.capacity()) {
      var minSizeIncrease = (buffer.capacity() * 3) / 2 + 1;
      var newBuffer = ByteBuffer.allocateDirect(Math.max(minSizeIncrease, buffer.capacity() + toAdd.capacity()));
      newBuffer.put(this.utf8());
      buffer = newBuffer;
    }
    buffer.put(toAdd.duplicate());
  }
}

File: assets/rt-js/MutStr.js
import { BaseStr } from "./BaseStr.js";
import { rt$$NativeRuntime } from "./NativeRuntime.js";
import { base$$True_0 } from "../base/True_0.js";
import { base$$False_0 } from "../base/False_0.js";
import { base$$Void_0 } from "../base/Void_0.js";
import { rt$$Str } from "./Str.js";

export class rt$$MutStr extends BaseStr {
  constructor(str) {
    super();
    // Start with an empty buffer, grow as needed
    this._buffer = new Uint8Array(16);
    this._length = 0; // actual bytes used
    this._graphemes = null;
    if (str) {
      this.put(str);
    }
  }

  // Return a Str-like object with the current bytes
  utf8() {
    return this._buffer.slice(0, this._length);
  }

  // Compute or return cached grapheme indices
  graphemes() {
    if (this._graphemes) return this._graphemes;
    this._graphemes = rt$$NativeRuntime.indexString(this.utf8());
    return this._graphemes;
  }

  // Is the string empty?
  isEmpty$read$0() {
    return this._length === 0 ? base$$True_0.$self : base$$False_0.$self;
  }

  // Mutable concatenation
  $plus$mut$1(other$) {
    if (typeof other$.str$read$0 === "function") {
      const other = other$.str$read$0();
      this.put(other);
    }
    return this;
  }

  append$mut(other$) {
    this.$plus$mut$1(other$);
    return base$$Void_0.$self;
  }

  clear$mut() {
    this._buffer = new Uint8Array(16);
    this._length = 0;
    this._graphemes = null;
    return base$$Void_0.$self;
  }

  str$read$0() {
    return rt$$Str.fromTrustedUtf8(this.utf8());
  }

  // --- private helper ---
  put(str) {
    this._graphemes = null;
    // const bytes = str.utf8();
    // Accept plain JS strings or rt$$Str
    let bytes;
    if (typeof str === "string") {
      bytes = new TextEncoder().encode(str);
    } else if (str && typeof str.utf8 === "function") {
      bytes = str.utf8();
    } else {
      throw new TypeError("MutStr.put: expected rt$$Str or string, got " + typeof str);
    }
    // Resize buffer if needed
    if (this._buffer.length - this._length < bytes.length) {
      const minSizeIncrease = Math.floor(this._buffer.length * 1.5) + 1;
      const newSize = Math.max(minSizeIncrease, this._buffer.length + bytes.length);
      const newBuffer = new Uint8Array(newSize);
      newBuffer.set(this._buffer.slice(0, this._length));
      this._buffer = newBuffer;
    }
    this._buffer.set(bytes, this._length);
    this._length += bytes.length;
  }

}
